algo/uidlist.go

```diff
@@ -105,46 +106,50 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
 	}
 
 	// Pick the shorter list and do binary search
-	if ld < lq {
+	if ld <= lq {
 		for {
 			blockUids := dec.Uids()
 			if len(blockUids) == 0 {
 				break
 			}
-			IntersectWithBin(blockUids, q, o)
-			lastUid := blockUids[len(blockUids)-1]
-			qidx := sort.Search(len(q), func(idx int) bool {
-				return q[idx] >= lastUid
-			})
-			if qidx >= len(q) {
+			if ld*linVsBinRatio < len(q) {
+				q = q[IntersectWithBin(blockUids, q, o):]
+			} else {
+				// For small enough difference between two arrays, we should just
+				// do lin intersect
+				_, off := IntersectWithLin(blockUids, q, o)
+				q = q[off:]
+			}
+			if len(q) == 0 {
 				return
 			}
-			q = q[qidx:]
 			dec.Next()
 		}
 		return
 	}
 
-	var uids []uint64
-	for _, u := range q {
+	uids := dec.Uids()
+	qidx := 0
+	for {
+		if qidx >= len(q) {
+			return
+		}
+		u := q[qidx]
 		if len(uids) == 0 || u > uids[len(uids)-1] {
-			uids = dec.Seek(u, codec.SeekStart)
+			if lq*linVsBinRatio < ld {
+				uids = dec.LinearSeek(u)
+			} else {
+				uids = dec.SeekToBlock(u, codec.SeekCurrent)
```

Comment:
I am not sure what the difference is between Seek & SeekToBlock here, and how it helps. Maybe add an explanation with some context.

Reply:
I have added that as a comment in SeekToBlock. Basically When we seek, we figure out where the uid is in the big block of uids that receive from decoder. We dont need to do this, as we are already going to do similar stuff while intersecting. Explanation is in comment for SeekToBlock

Reply:
// SeekToBlock will find the nearest block, and unpack it. Unlike Seek, it doesn't
// apply search in the resulting uid list and then move the pointer forward. When we are going
// to intersect the list later, this function is useful.